domain/application/state/unit.go

```diff
@@ -559,6 +561,100 @@ func (st *State) InsertMigratingCAASUnits(ctx context.Context, appUUID coreappli
 	})
 }
 
+// SetRunningAgentBinaryVersion sets the running agent binary version for the
+// provided unit uuid. Any previously set values for this unit uuid will
+// be overwritten by this call.
+//
+// The following errors can be expected:
+// - [errors.UnitNotFound] if the unit does not exist.
+// - [coreerrors.NotSupported] if the architecture is not known to the database.
+func (st *State) SetRunningAgentBinaryVersion(ctx context.Context, uuid coreunit.UUID, version coreagentbinary.Version) error {
+	db, err := st.DB()
+	if err != nil {
+		return errors.Capture(err)
+	}
+
+	type ArchitectureMap struct {
+		ID   int    `db:"id"`
+		Name string `db:"name"`
+	}
+	archMap := ArchitectureMap{Name: version.Arch}
+
+	archMapStmt, err := st.Prepare(`
+SELECT id AS &ArchitectureMap.id FROM architecture WHERE name = $ArchitectureMap.name
+`, archMap)
+	if err != nil {
+		return errors.Capture(err)
+	}
+
+	unitUUID := unitUUID{UnitUUID: uuid}
+
+	type UnitAgentVersion struct {
+		UnitUUID      string `db:"unit_uuid"`
+		Version       string `db:"version"`
+		ArchtectureID int    `db:"architecture_id"`
+	}
+	unitAgentVersion := UnitAgentVersion{
+		UnitUUID: unitUUID.UnitUUID.String(),
+		Version:  version.Number.String(),
+	}
+
+	upsertRunningVersionStmt, err := st.Prepare(`
+INSERT INTO unit_agent_version (*)
+VALUES ($UnitAgentVersion.*)
+ON CONFLICT (unit_uuid) DO
+UPDATE SET version = excluded.version,
+           architecture_id = excluded.architecture_id
+`, unitAgentVersion)
+	if err != nil {
+		return errors.Capture(err)
+	}
+
+	err = db.Txn(ctx, func(ctx context.Context, tx *sqlair.TX) error {
+
+		// Check if unit exists
+		err := st.checkUnitNotDead(ctx, tx, unitUUID)
+		switch {
+		case errors.Is(err, applicationerrors.UnitNotFound):
+			return errors.Errorf(
+				"unit %q does not exist", unitUUID,
+			).Add(applicationerrors.UnitNotFound)
+		case errors.Is(err, applicationerrors.UnitIsDead):
+			return errors.Errorf(
+				"unit %q is dead", unitUUID,
+			).Add(applicationerrors.UnitIsDead)
+		case err != nil:
+			return errors.Errorf(
+				"checking unit %q exists: %w", unitUUID, err,
+			)
+		}
```

Comment:
The `checkUnitNotDead` method returns the correct error type already. The wording needs improvement perhaps, but just use the same as is done already elsewhere, ie
```
		err := st.checkUnitNotDead(ctx, tx, unitUUID)
		if err != nil {
			return errors.Errorf("checking unit %q exists: %w", uuid, err)
		}
```

Reply:
fixed.